### ISSUE TITLE
Feature Constructor loses features when input changed

### ISSUE THREAD

**Issue by Mark432432** [biolab/orange3#4415]
I have a FILE widget followed by a SELECT COLUMNS followed by a FEATURE CONSTRUCTOR

I have created 20 or so features in FEATURE CONSTRUCTOR..

When in the SELECT COLUMNS if I add an additional variable to the FEATURES list I lose all but 2 of my constructed features when I open the FEATURE CONSTRUCTOR widget.

**To Reproduce**
I am not yet sure if this is reproducible as I have a bit of work to reproduce.  I will probably create a separate flow to add more features.

**Orange version:**
Orange - About - 3.24.1

**Expected behavior**
To not lose any features in FEATURE CONSTRUCTOR when it's input is changed.

**Screenshots**

**Operating system:**
Windows, OSX or Linux? Which version? Windows 10

**Additional context**

**Comment by janezd** [comment]
Do you by any chance have this schema saved? Can you upload an .ows file?

I have a hypothesis but it's difficult to check.

**Comment by Mark432432** [comment]
[train.zip](https://github.com/biolab/orange3/files/4196138/train.zip)

The error occurred when I add "YrSold" to the features in "Select Column - Scale".

**Comment by Mark432432** [comment]
Hi,
I updated the github issue.

Let me know if you need anything else.
Thanks,
Mark

    On Wednesday, February 12, 2020, 4:18:14 PM EST, Janez Demšar <notifications@github.com> wrote:  
 
 
Do you by any chance have this schema saved? Can you upload an .ows file?

I have a hypothesis but it's difficult to check.

**Comment by ajdapretnar** [comment]
I can't replicate this based on your instructions. The FeatConst stays the same.

**Comment by janezd** [comment]
I would need an .ows more than the data.

Say that a widget receives some data, and you change the widget's settings somehow, e.g. select x and y. 

- Next the widget gets an empty input (perhaps the input comes from some selection in another widget). It has to reset to defaults (that is, empty).
- Next it again receives some new data sample, but apparently from the same data because it still has the same attributes x, y and color that were used previously. Of course you would expect the widget to reset to its previous settings. And it does.
- Now let this widget get some completely different data. It has to reset to some defaults.
- Now let this widget again receive the same data as before. It would be nice if it remembered the settings from the history. It does.
- Now let this widget receive similar data, but with some variables missing. Maybe you removed them in Select Columns. It can (and will) retain the settings if the attributes used for x and y are present. For the sake of example, let us say that you also change something else, for instance the color attribute.

At this point, the widget remembers three different sets of settings: one is for the initial data, one is for the completely different second data and the last one is for the last data which was compatible, but different from the first one.

Finally, the widget receives some fourth data set, which is "compatible" with the first and the third. Which settings should it retrieve? I would say that those which are the most similar to the data. Moreover, if one of the saved settings have exactly the same attributes, it takes if be default.

I believe that in your case Orange works as intended, but this behaviour sometimes seems incorrect. You probably first sent the entire data to feature constructor. You constructed no features. Next you removed some attributes. You constructed some features. When you add attributes, the data becomes more similar to the initial data, which contained no features.

We might be able to see this from .ows (but not necessarily). Or my hypothesis may be wrong altogether.

We have encountered this in the past, I believe it was related to Select Rows. Behaviour of Orange could be changed so that it grabs **the most recent** compatible context, ignoring how similar it is. I however fear this would look wrong in other scenarious.

**Comment by Mark432432** [comment]
Hi,
I thought I attached a zip file with a data file (train.csv) and an .ows file?
One other note, after I update the 'select columns' 2 of the 20+ features do remain in the feature constructor.  Not exactly sure why these 2 remain.
it is probable that at one point I did have all of the file attributes passing thru the select columns and into the feature constructor.
I just checked github and I do see 2 files in the .zip one is the .ows.
Thanks,
Mark



    On Thursday, February 13, 2020, 4:18:38 AM EST, Janez Demšar <notifications@github.com> wrote:  
 
 
I would need an .ows more than the data.

Say that a widget receives some data, and you change the widget's settings somehow, e.g. select x and y.
   
   - Next the widget gets an empty input (perhaps the input comes from some selection in another widget). It has to reset to defaults (that is, empty).
   - Next it again receives some new data sample, but apparently from the same data because it still has the same attributes x, y and color that were used previously. Of course you would expect the widget to reset to its previous settings. And it does.
   - Now let this widget get some completely different data. It has to reset to some defaults.
   - Now let this widget again receive the same data as before. It would be nice if it remembered the settings from the history. It does.
   - Now let this widget receive similar data, but with some variables missing. Maybe you removed them in Select Columns. It can (and will) retain the settings if the attributes used for x and y are present. For the sake of example, let us say that you also change something else, for instance the color attribute.

At this point, the widget remembers three different sets of settings: one is for the initial data, one is for the completely different second data and the last one is for the last data which was compatible, but different from the first one.

Finally, the widget receives some fourth data set, which is "compatible" with the first and the third. Which settings should it retrieve? I would say that those which are the most similar to the data. Moreover, if one of the saved settings have exactly the same attributes, it takes if be default.

I believe that in your case Orange works as intended, but this behaviour sometimes seems incorrect. You probably first sent the entire data to feature constructor. You constructed no features. Next you removed some attributes. You constructed some features. When you add attributes, the data becomes more similar to the initial data, which contained no features.

We might be able to see this from .ows (but not necessarily). Or my hypothesis may be wrong altogether.

We have encountered this in the past, I believe it was related to Select Rows. Behaviour of Orange could be changed so that it grabs the most recent compatible context, ignoring how similar it is. I however fear this would look wrong in other scenarious.

**Comment by janezd** [comment]
> I thought I attached a zip file with a data file (train.csv) and an .ows file?

You have. Sorry.

My hypothesis is unfortunatelly correct. If your data contains variables

```
{'1stFlrSF': 2, 'LotFrontage': 2, 'LotArea': 2, 'MasVnrArea': 2, 'BsmtFinSF1': 2, 'BsmtFinSF2': 2, 'BsmtUnfSF': 2, 'TotalBsmtSF': 2, '2ndFlrSF': 2, 'LowQualFinSF': 2, 'GrLivArea': 2, 'BsmtFullBath': 2, 'BsmtHalfBath': 2, 'FullBath': 2, 'HalfBath': 2, 'BedroomAbvGr': 2, 'KitchenAbvGr': 2, 'TotRmsAbvGrd': 2, 'Fireplaces': 2, 'GarageYrBlt': 2, 'GarageCars': 2, 'GarageArea': 2, 'WoodDeckSF': 2, 'OpenPorchSF': 2, 'EnclosedPorch': 2, '3SsnPorch': 2, 'ScreenPorch': 2, 'PoolArea': 2, 'MiscVal': 2, 'OverallQual': 2, 'OverallCond': 2, 'SalePrice': 2}
```

widget will contain the following expressions:

```
[ContinuousDescriptor(name='1stFlrSF_Scaled', expression='_1stFlrSF*10/4692', number_of_decimals=3), ContinuousDescriptor(name='2ndFlrSF_Scaled', expression='_2ndFlrSF*10/2065', number_of_decimals=3), ContinuousDescriptor(name='3SsnPorch_Scaled', expression='_3SsnPorch*10/508', number_of_decimals=3), ContinuousDescriptor(name='BedroomAbvGr_Scaled', expression='BedroomAbvGr*10/8', number_of_decimals=3), ContinuousDescriptor(name='BSMTFinSF1_Scalled', expression='BsmtFinSF1*10/5644', number_of_decimals=3), ContinuousDescriptor(name='BSMTFinSF2_Scalled', expression='BsmtFinSF2*10/1474', number_of_decimals=3), ContinuousDescriptor(name='BsmtFullBath_Scaled', expression='BsmtFullBath*10/3', number_of_decimals=3), ContinuousDescriptor(name='BsmtHalfBath_Scaled', expression='BsmtHalfBath*10/2', number_of_decimals=3), ContinuousDescriptor(name='BsmtUnfSF_Scaled', expression='BsmtUnfSF*10/2336', number_of_decimals=3), ContinuousDescriptor(name='EnclosedPorch_Scaled', expression='EnclosedPorch*10/552', number_of_decimals=3), ContinuousDescriptor(name='Fireplaces_Scaled', expression='Fireplaces*10/3', number_of_decimals=3), ContinuousDescriptor(name='FullBath_Scaled', expression='FullBath*10/3', number_of_decimals=3), ContinuousDescriptor(name='GarageArea_Scaled', expression='GarageArea*10/1418', number_of_decimals=3), ContinuousDescriptor(name='GarageCars_Scaled', expression='GarageCars*10/4', number_of_decimals=3), ContinuousDescriptor(name='GrLivArea_Scaled', expression='GrLivArea*10/5642', number_of_decimals=3), ContinuousDescriptor(name='HalfBath_Scaled', expression='HalfBath*10/2', number_of_decimals=3), ContinuousDescriptor(name='KitchenAbvGr_Scale', expression='KitchenAbvGr*10/3', number_of_decimals=3), ContinuousDescriptor(name='LotArea_Scaled', expression='LotArea*10/215245', number_of_decimals=3), ContinuousDescriptor(name='LotFrontage_Scaled', expression='LotFrontage*10/313', number_of_decimals=3), ContinuousDescriptor(name='LowQualFinSF_Scaled', expression='LowQualFinSF*10/572', number_of_decimals=3), ContinuousDescriptor(name='MasVnrArea_Scaled', expression='MasVnrArea*10/1600', number_of_decimals=3), ContinuousDescriptor(name='MiscVal_Scaled', expression='MiscVal*10/15500', number_of_decimals=3), ContinuousDescriptor(name='OpenPorchSF_Scaled', expression='OpenPorchSF*10/547', number_of_decimals=3), ContinuousDescriptor(name='OverallCond_Scaled', expression='OverallCond', number_of_decimals=3), ContinuousDescriptor(name='PoolArea_Scaled', expression='PoolArea*10/738', number_of_decimals=3), ContinuousDescriptor(name='ScreenPorch_Scaled', expression='ScreenPorch*10/480', number_of_decimals=3), ContinuousDescriptor(name='TotalBsmtSF_Scaled', expression='TotalBsmtSF*10/6110', number_of_decimals=3), ContinuousDescriptor(name='TotalRmsAbvGrd_Scaled', expression='TotRmsAbvGrd*10/14', number_of_decimals=3), ContinuousDescriptor(name='WoodDeckSF_Scaled', expression='WoodDeckSF*10/857', number_of_decimals=3)], '__version__': 1})
```

However, if you have

```
{'LotFrontage': 2, 'LotArea': 2, 'MasVnrArea': 2, 'BsmtFinSF1': 2, 'BsmtFinSF2': 2, 'BsmtUnfSF': 2, 'TotalBsmtSF': 2, '1stFlrSF': 2, '2ndFlrSF': 2, 'LowQualFinSF': 2, 'GrLivArea': 2, 'BsmtFullBath': 2, 'BsmtHalfBath': 2, 'FullBath': 2, 'HalfBath': 2, 'BedroomAbvGr': 2, 'KitchenAbvGr': 2, 'TotRmsAbvGrd': 2, 'Fireplaces': 2, 'GarageYrBlt': 2, 'GarageCars': 2, 'GarageArea': 2, 'WoodDeckSF': 2, 'OpenPorchSF': 2, 'EnclosedPorch': 2, '3SsnPorch': 2, 'ScreenPorch': 2, 'PoolArea': 2, 'MiscVal': 2, 'SalePrice': 2}
```

you have just

```
[ContinuousDescriptor(name='1stFlrSF_scaled', expression='_1stFlrSF*10/5095', number_of_decimals=3), ContinuousDescriptor(name='2ndFlrSF_Scaled', expression='_2ndFlrSF*10/2065', number_of_decimals=3)]
```

**Comment by janezd** [comment]
To replicate:
- File -> Select Columns -> Feature Constructor.
- In Feature Constructor, add expression X1 = 42.
- In Select Columns, remove the first attribute. **X1 = 42 remains in Feature constructor.**
- In Feature Constructor, add X2 = 42
- In Select Columns, add the first attribute. **X2 = 42 disappears.**

(Adding X1 = 42 is not needed, but it's good to check that it remains after removing an attribute.)

This is not strictly a bug -- Orange works as intended, but perhaps it could work differently. To be discussed.

**Comment by janezd** [comment]
Result of discussion: we decided to modify Orange so that it will use the latest compatible context.

**Comment by Mark432432** [comment]
Thank you!.
Any way to avoid this, until the change?
I am a little concerned about losing extensive transformations if the input to the feature constructor changes.
I guess I could move all file attributes into the select columns from the start, and then hopefully not have to add any columns to the file.
Thanks,
Mark



    On Friday, February 14, 2020, 5:27:59 AM EST, Janez Demšar <notifications@github.com> wrote:  
 
 
Result of discussion: we decided to modify Orange so that it will use the latest compatible context.

**Comment by janezd** [comment]
> I guess I could move all file attributes into the select columns from the start, and then hopefully not have to add any columns to the file.

Yes, this seems to be the only solution at the moment. If everything goes well, the fix will be included in next release of Orange.

**Comment by janezd** [comment]
Fixed via #4421 and https://github.com/biolab/orange-widget-base/pull/42.
